src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -103,21 +104,37 @@ public ImmutableArray<PackageSource> GetPackageSources()
                 return packageSources;
             }
 
-            try
+            if (_packageSourcesAsync is null)
             {
-                packageSources = _packageSourceProvider.Value.GetSources(includeUnOfficial: true, includeDisabled: false)
-                    .SelectAsArray(r => new PackageSource(r.Key, r.Value));
+                lock (_gate)
+                {
+                    if (_packageSourcesAsync is null)
+                    {
+                        _packageSourcesAsync = ThreadingContext.JoinableTaskFactory.RunAsync(() => GetPackageSourcesImplAsync());
+                    }
+                }
             }
-            catch (Exception ex) when (ex is InvalidDataException || ex is InvalidOperationException)
+
+            if (allowSwitchToMainThread)
             {
-                // These exceptions can happen when the nuget.config file is broken.
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                using var combinedToken = _tokenSource.Token.CombineWith(cancellationToken);
+                packageSources = await _packageSourcesAsync.JoinAsync(combinedToken.Token).ConfigureAwait(false);
```

Comment:
why does this end up switching to the main thread?

Reply:
STA COM